### ISSUE TITLE
4.0.0 Fehlerhafte defaults-Werte in Table-Manager-Tabelle bei Textfeldern (führen zu Problemen)

### ISSUE THREAD

**Issue by alxndr-w** [yakamara/yform#1215]
Erläuterung: MariaDB ist in der Lage, defaults-Werte zu speichern für Felder. Diese werden über den Table Manager in `rex_yform_field` definiert, hier am Beispiel der Tabelle `rex_ycom_user`. 

Standardmäßig ist der default-Wert für Textfelder leer und nicht nullable. 

Beim Update der Felder werden vom Table Manager dann fälschlicherweise statt einem leeren String Anführungszeichen `''` eingetragen.

gelb: So, wie YForm standardmäßig Felder einträgt.
rote Häkchen: So, wie ich das Problem aktuell umgehe mit `NULL`

![image](https://user-images.githubusercontent.com/3855487/147882319-60b12351-ebea-4c5c-96ac-7af0564c072c.png)

gelb: Das, was YForm als default-Wert einträgt und zu Problemen führt.

![image](https://user-images.githubusercontent.com/3855487/147882308-6cbb00bc-6557-4b9f-821a-9de9eb5877b9.png)

Das führt aktuell bei einem Projekt mit @tyrant88 am Ende zu Problemen bei YDeploy, wenn lokal mit MariaDB und live mit MySQL gearbeitet wird, das scheinbar nicht in der Lage ist, die default-Werte zu verwenden.

```

    rex_sql_table::get('rex_ycom_user')
        ->ensureColumn(new rex_sql_column('salutation', 'text', false, null, null), 'ycom_groups')
        ->ensureColumn(new rex_sql_column('firstname', 'varchar(191)', false, '', null))
        ->ensureColumn(new rex_sql_column('lastname', 'varchar(191)', false, '', null), 'firstname')
        ->ensureColumn(new rex_sql_column('contact_name_show', 'text', false, '\'\'', null), 'lastname') // <---- hier problematische default-Werte 
        ->ensureColumn(new rex_sql_column('email', 'varchar(191)', false, '', null), 'contact_name_show')
        ->ensureColumn(new rex_sql_column('contact_mail_show', 'text', false, '\'\'', null), 'email') // <---- hier problematische default-Werte 
        ->ensureColumn(new rex_sql_column('address', 'text', false, '\'\'', null), 'contact_mail_show') // <---- hier problematische default-Werte 
        ->ensureColumn(new rex_sql_column('address_show', 'tinyint(1)', false, '0', null), 'address')
        ->ensureColumn(new rex_sql_column('contact_show', 'text', false, '\'\'', null), 'address_show') // <---- hier problematische default-Werte 
        ->ensureColumn(new rex_sql_column('login', 'varchar(191)', false, '', null), 'contact_show')
        ->ensureColumn(new rex_sql_column('password', 'varchar(255)', false, null, null), 'login')
        ->ensureColumn(new rex_sql_column('status', 'text', false, null, null), 'password')
        ->ensureColumn(new rex_sql_column('activation_key', 'varchar(191)', false, null, null), 'status')
        ->ensureColumn(new rex_sql_column('session_key', 'varchar(191)', false, null, null), 'activation_key')
        ->ensureColumn(new rex_sql_column('termsofuse_accepted', 'tinyint(1)', false, '0', null), 'session_key')
        ->ensureColumn(new rex_sql_column('new_password_required', 'tinyint(1)', false, '0', null), 'termsofuse_accepted')
        ->ensureColumn(new rex_sql_column('last_action_time', 'datetime', false, null, null), 'new_password_required')
        ->ensureColumn(new rex_sql_column('last_login_time', 'datetime', false, null, null), 'last_action_time')
        ->ensureColumn(new rex_sql_column('termination_time', 'datetime', false, null, null), 'last_login_time')
        ->ensureColumn(new rex_sql_column('login_tries', 'int(11)', true, null, null), 'termination_time')
        ->ensureColumn(new rex_sql_column('name', 'varchar(191)', false, '', null))
        ->alter();
```

Standard-Wert sollte `NULL` sein oder gar nicht gesetzt werden (`kein(e)`) damit 1. der default überhaupt Wert korrekt ist (leer, nicht die Zeichen `''`), 2. das einem nicht um die Ohren fliegt, wenn man einen DBMS-Wechsel hat.

**Comment by gharlan** [comment]
Verwandtes Issue: https://github.com/yakamara/ydeploy/issues/39
Und: https://github.com/redaxo/redaxo/pull/3270

**Comment by tyrant88** [comment]
Okay, danke für die ganzen Infos!
Also muss der, der Maria-Db nutzt aufpassen, wenn er neue Felder anlegt... :-)

**Comment by gharlan** [comment]
Ja, wobei somit yform am besten auch gar nicht erst versuchen sollte, für TEXT-Spalten Default-Werte zu setzen.

**Comment by tyrant88** [comment]
Ja, klar, ich gehe davon aus, dass yform entsprechend gefixt wird. 
Jan hat es ja schon dem 4.0 Milestone zugeordnet ;-)
Und das müsste jedes AddOn machen, das TEXT-Felder in der Datenbank anlegt, richtig?

**Comment by gharlan** [comment]
Ja. Wobei mir aktuell keine weiteren problematischen Fälle bekannt sind.
Ist das Problem in yform eigentlich neu in 4.0? Wisst ihr das?
Denn wundere mich, dass es bei uns noch nicht aufgetreten ist. Wir arbeiten lokal mit mariadb, produktiv haben wir aber öfters mysql. So war ja auch der metainfo-Fall bei uns aufgefallen.

**Comment by tyrant88** [comment]
Bei uns traten die Probleme erst mit dem Update auf 4.0.0 auf.

**Comment by alxndr-w** [comment]
Was mir eben sofort ins Auge gesprungen ist, ist, dass escapte Anführungszeichen als Wert verwendet werden. Also als ob irgendwo in einer Methode statt ` ` ein `''` übergeben wird (leerer escapter String?) und daraus dann `\'\'` wird.

Ein leerer String wäre ja auch ebenfalls korrekt. Nur eben nicht `''` als Default-Wert.

**Comment by gharlan** [comment]
MariaDB liefert bei TEXT-Columns so die Default-Values. Also mit zusätzlichen Anführungszeichen. Warum das so ist, weiß ich gerade nicht (mehr).
Und die Backslashes kommen halt hinzu, weil der String als PHP-String mit einfachen Anführungszeichen in der Datei steht. Letztlich ist es ein String mit zwei Anführungszeichen drin, ohne Backslashes.

Ich bin mir gerade nicht ganz sicher, meine aber, dass innerhalb von MariaDB das so auch dann wieder funktioniert, wenn der Default-Wert so gesetzt wird. Bloß in Kombination mit MySQL funktioniert es so nicht. MySQL unterstützt allgemein keine Default-Values für TEXT-Columns, wirft aber anscheinend nur Fehler bei diesen komischen Values aus MariaDB mit zusätzlichen Anführungszeichen.

So ganz genau weiß ich es nicht mehr alles. Aber im Metainfo-Fall konnten wir es halt lösen, indem wir gar nicht mehr versuchen, Default-Values für TEXT-Columns zu setzen, da MySQL sie halt sowieso nicht unterstützt.

**Comment by alxndr-w** [comment]
Das kommt mir spanisch vor. Dann müsste man das ja schon bei dem Abfrage-Ergebnis an MariaDB für sich zu einem leeren String oder null abwandeln und schon wäre das Problem gelöst, oder?

**Comment by gharlan** [comment]
MariaDB liefert die Anführungszeichen, weil auch andere Expressions möglich sind als Default-Value, wie `CONCAT(...)`.
Deswegen kann es problematisch sein, die Anführungszeichen zu entfernen. Speziell für den leeren String könnte man es vermulich aber machen.

**Comment by alxndr-w** [comment]
Ich lebe jetzt einfach weiter damit, in jedem Deployment erstmal einen Fehler zu bekommen, bis ich es lerne, nach jedem migrate mit "suchen und Ersetzen" die Datei zu korrigieren. Ist ja eh kein originärer Fehler von YForm.
